vendor/src/github.com/docker/libnetwork/datastore/datastore.go

```diff
@@ -283,11 +305,31 @@ func (ds *datastore) Watch(kvObject KVObject, stopCh <-chan struct{}) (<-chan KV
 				kvoCh <- dstO
 			}
 		}
+
+		// Wait on watch channel for a re-trigger when datastore becomes active
+		<-watchCh
+
+		kvpCh, err = ds.store.Watch(Key(kvObject.Key()...), sCh)
+		if err != nil {
+			log.Printf("Could not watch the key %s in store: %v", Key(kvObject.Key()...), err)
+		}
+
+		goto retry_watch
```

Comment:
@abronan we can certainly talk about splitting it into multiple functions (which I have to see how best to organize it). But the watch logic is actually an interesting golang's channel pattern for 1->many receivers. Once you understand the intention, you can appreciate @mrjana's idea on managing it this way. So, let me explain the logic a bit. The only watch that libnetwork has today is the network watch (not a watchtree) for SD. Since libkv doesn't (and cannot) provide a retry logic for watches, its the applications responsibility to do retrigger the watch when the store restarts. we need a trigger to restart the watch for all the created and active networks when the store is back online. Instead of writing a registration mechanism for a broadcast channel pattern, we used a simple trick for this notification. By default if there are multiple receivers for a channel and there is no data sent on that channel, they will all be in blocked state. When the channel is closed, every blocked receiver will be notified. We use this pattern that reduces the code-complexity greatly. Since we are not looking for any actual data on the receivers but just a trigger to restart the watch, this fits perfectly well. Also, as you can see, we dont even do a periodic retry to connect to the data-store. that also helps to reduce the code complexity greatly.